### PR TITLE
Master stock val cur rate reconcile oco

### PR DESCRIPTION
**Pull request by oco-odoo**
In anglosaxon accounting, perform reconciliation on stock interim account between stock valuation journal items and invoice's journal items.

### CODE REVIEW

File: addons/purchase/models/account_invoice.py [hunk 1]
```diff
@@ -209,6 +225,24 @@ def write(self, vals):
...
         return result
 
+    def _get_related_stock_moves(self):
+        """ Overridden from stock_account.
+        Returns the stock moves associated to this invoice."""
+        rslt = super(AccountInvoice, self)._get_related_stock_moves()
+
+        if self.type == 'in_invoice':
+            for inv_line in self.invoice_line_ids:
+                rslt += inv_line.purchase_line_id.move_ids
```
Comment (xmo-odoo) [inline addons/purchase/models/account_invoice.py]: Why not use `self.mapped('invoice_line_ids.purchase_line.move_ids')`?

File: addons/sale_stock/models/account_invoice.py [hunk 2]
```diff
@@ -11,6 +11,21 @@ class AccountInvoice(models.Model):
...
         readonly=True, states={'draft': [('readonly', False)]})
 
+    def _get_related_stock_moves(self): # overridden from stock_account
+        rslt = super(AccountInvoice, self)._get_related_stock_moves()
+
+        if self.type == 'out_invoice':
+            for inv_line in self.invoice_line_ids:
+                for so_line in inv_line.sale_line_ids:
+                    for picking in so_line.order_id.picking_ids:
+                        rslt += picking.move_lines
```
Comment (xmo-odoo) [inline addons/sale_stock/models/account_invoice.py]: Or just `self.invoice_line_ids.sale_line_ids.order_id.picking_ids`.

File: addons/stock_account/models/account_invoice.py [hunk 3]
```diff
@@ -77,6 +77,52 @@ def _anglo_saxon_sale_move_lines(self, i_line):
...
+        return the stock moves related to this invoice.
+        """
+        return self.env['stock.move']
+
+    def _get_products_set(self):
+        """ Returns a recordset of the products contained in this invoice's lines
+        """
+        rslt = self.env['product.product']
+        for inv_line in self.invoice_line_ids:
+            rslt += inv_line.product_id
```
Comment (xmo-odoo) [inline addons/stock_account/models/account_invoice.py]: `return self.invoice_line_ids.product_id`?

File: addons/stock_account/models/account_invoice.py [hunk 4]
```diff
@@ -77,6 +77,52 @@ def _anglo_saxon_sale_move_lines(self, i_line):
...
+        return rslt
+
+    def _get_anglosaxon_interim_account(self, product): #
+        """ To be overriddent for customer invoices and vendor bills in order to
```
Comment (xmo-odoo) [inline addons/stock_account/models/account_invoice.py]: Typo

File: addons/stock_account/models/account_invoice.py [hunk 5]
```diff
@@ -77,6 +77,52 @@ def _anglo_saxon_sale_move_lines(self, i_line):
...
+        super(AccountInvoice, self).invoice_validate()
+        self.anglo_saxon_reconcile_valuation()
+
+    def anglo_saxon_reconcile_valuation(self):
+        """ Reconciles the entries made in the interim accounts in anglosaxon accounting,
+        reconciling stock valuation move lines with the invoice's.
+        """
+        for invoice in self:
+            if invoice.company_id.anglo_saxon_accounting:
+                invoice_stock_moves_id_list = [move.id for move in self._get_related_stock_moves()]
```
Comment (xmo-odoo) [inline addons/stock_account/models/account_invoice.py]: `self._get_related_stock_moves().ids`?

File: addons/stock_account/models/account_invoice.py [hunk 6]
```diff
@@ -77,6 +77,52 @@ def _anglo_saxon_sale_move_lines(self, i_line):
...
+                    if product.valuation == 'real_time' and product.cost_method == 'real':
+                        # We first get the invoice's move lines ...
+                        product_interim_account = invoice._get_anglosaxon_interim_account(product)
+                        to_reconcile = self.env['account.move.line'].search([('move_id','=',invoice.move_id.id), ('product_id','=',product.id), ('account_id','=',product_interim_account.id), ('reconciled','=',False)])
+
+                        # And then the stock valuation ones.
+                        product_stock_moves = self.env['stock.move'].search([('id','in',invoice_stock_moves_id_list), ('product_id','=',product.id)])
+                        for stock_move in product_stock_moves:
+                            for valuation_move in stock_move.stock_account_valuation_account_move_ids:
+                                for valuation_line in valuation_move.line_ids:
```
Comment (xmo-odoo) [inline addons/stock_account/models/account_invoice.py]: Can be folded in a single mapped call.

File: addons/stock_account/tests/#__init__.py# [hunk 7]
```diff
@@ -0,0 +1,4 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from . import test_valuation_reconciliation_common
```
Comment (xmo-odoo) [inline addons/stock_account/tests/#__init__.py#]: crap file

File: addons/stock_account/models/account_invoice.py [hunk 8]
```diff
@@ -38,7 +38,7 @@ def _anglo_saxon_sale_move_lines(self, i_line):
             cacc = accounts['expense'].id
             if dacc and cacc:
                 price_unit = i_line._get_anglo_saxon_price_unit()
-                if inv.currency_id.id != company_currency:
+                if inv.currency_id != company_currency:
```
Comment (qdp-odoo) [inline addons/stock_account/models/account_invoice.py]: that bugfix should be commited in stable, it has nothing to do in this branch

File: addons/purchase/models/account_invoice.py [hunk 9]
```diff
@@ -142,6 +145,8 @@ def _anglo_saxon_purchase_move_lines(self, i_line, res):
             for line in res:
                 if line.get('invl_id', 0) == i_line.id and reference_account_id == line['account_id']:
                     valuation_price_unit = i_line.product_id.uom_id._compute_price(i_line.product_id.standard_price, i_line.uom_id)
+                    interim_account_price = float_round(valuation_price_unit * line['quantity'], precision_digits=account_prec)
```
Comment (qdp-odoo) [inline addons/purchase/models/account_invoice.py]: if there is a bug, please make a PR to the proper stable version where it ocured first, with use case explained and so on

File: addons/purchase/models/account_invoice.py [hunk 10]
```diff
@@ -209,6 +225,24 @@ def write(self, vals):
                 invoice.message_post(body=message)
         return result
 
+    def _get_related_stock_moves(self):
+        """ Overridden from stock_account.
+        Returns the stock moves associated to this invoice."""
+        rslt = super(AccountInvoice, self)._get_related_stock_moves()
+
+        if self.type == 'in_invoice':
```
Comment (qdp-odoo) [inline addons/purchase/models/account_invoice.py]:
```python
if self.type in ('in_invoice', ('in_refund'):
```

Comment (qdp-odoo) [inline addons/purchase/models/account_invoice.py]: I suppose... need to test how refunds are treated in purchase module

File: addons/sale_stock/models/account_invoice.py [hunk 11]
```diff
@@ -11,6 +11,21 @@ class AccountInvoice(models.Model):
         help="Incoterms are series of sales terms. They are used to divide transaction costs and responsibilities between buyer and seller and reflect state-of-the-art transportation practices.",
         readonly=True, states={'draft': [('readonly', False)]})
 
+    def _get_related_stock_moves(self): # overridden from stock_account
+        rslt = super(AccountInvoice, self)._get_related_stock_moves()
+
+        if self.type == 'out_invoice':
```
Comment (qdp-odoo) [inline addons/sale_stock/models/account_invoice.py]:
```python
if self.type in ('out_invoice', 'out_refund'):
```
I suppose... need to test how refunds are treated in sale module

File: addons/stock_account/models/account_invoice.py (hunk 8 above)
Comment (oco-odoo) [inline addons/stock_account/models/account_invoice.py]: Done in PR https://github.com/odoo/odoo/pull/17259

File: addons/purchase/models/account_invoice.py (hunk 9 above)
Comment (oco-odoo) [inline addons/purchase/models/account_invoice.py]: Done for 9.0 in PR https://github.com/odoo/odoo/pull/17270

**Pull request closed without merging**